src/thorlabs_mff_fastcs/controllers.py

```diff
@@ -58,6 +61,56 @@ def read_hardware_v(self, response: bytes) -> int:
 protocol = ThorlabsAPTProtocol()
 
 
+@dataclass
+class ResponseCache:
+    _last_update: datetime | None = None
+    _response: str | None = None
+    _update_event: asyncio.Event = asyncio.Event()
+
+    def __post_init__(self) -> None:
+        self._update_event.set()
+
+    def check_expired(self, time_step: float) -> bool:
+        if self._last_update is None:
+            return True
+        delta_t = datetime.now() - self._last_update
+        return delta_t.total_seconds() > time_step
+
+    def update_response(self, response: str) -> None:
+        self._response = response
+        self._last_update = datetime.now()
+
+    async def get_response(
+        self, expiry_period: float, to_await: Callable[[Any], Awaitable[str]], *args
+    ):
+        read_cache = False
+
+        # Immediately short circuit if expired
+        if self.check_expired(expiry_period):
+            # Wait if an update is in progress
+            if self._update_event.is_set():
+                self._update_event.clear()
+            else:
+                await self._update_event.wait()
+                # Check if updated since
+                if not self.check_expired(expiry_period):
+                    read_cache = True
```

Comment:
I am not sure it is possible for this to still be expired after waiting on the update, is it? The task loop doing the update will block all the others here until it is complete.

Reply:
Discussed, agree it is unlikely - Will simplify code